plugins/tiledb/io/TileDBReader.cpp

```diff
@@ -0,0 +1,331 @@
+}
+
+void TileDBReader::addArgs(ProgramArgs& args)
+{
+    args.add("array_name", "TileDB array name", m_arrayName).setPositional();
+    args.add("config_file", "TileDB configuration file location", m_cfgFileName);
+    args.add("chunk_size", "TileDB read chunk size", m_chunkSize, point_count_t(1000000));
+    args.add("stats", "Dump TileDB query stats to stdout", m_stats, false);
+    args.add("bbox3d", "Bounding box subarray to read from TileDB in format "
+                       "([minx, maxx], [miny, maxy], [minz, maxz])", m_bbox);
+}
+
+void TileDBReader::initialize()
+{
+    if (!m_cfgFileName.empty())
+    {
+        tiledb::Config cfg(m_cfgFileName);
+        m_ctx.reset(new tiledb::Context(cfg));
+    }
+    else
+        m_ctx.reset(new tiledb::Context());
+
+    m_array.reset(new tiledb::Array(*m_ctx, m_arrayName, TILEDB_READ));
+}
+
+void TileDBReader::addDimensions(PointLayoutPtr layout)
+{
+    std::vector<tiledb::Dimension> dims = m_array->schema().domain().dimensions();
+    auto attrs = m_array->schema().attributes();
+
+    for (const auto& dim: dims)
+    {
+        layout->registerDim(Dimension::id(dim.name()));
+        m_ranges.push_back(dim.domain<double>().first);
+        m_ranges.push_back(dim.domain<double>().second);
+    }
+
+
+    for (const auto& a : attrs) {
+        std::string attrName = a.first;
+        Dimension::Id id = Dimension::id(attrName);
+        layout->registerDim(id);
+    }
+
+}
+
+point_count_t TileDBReader::read(PointViewPtr view, point_count_t count)
+{
+    PointId idx = view->size();
+    PointRef point(*view, idx);
+    point_count_t numRead = 0;
+    tiledb::Query query(*m_ctx, *m_array);
+    std::vector<tiledb::Dimension> dims = m_array->schema().domain().dimensions();
+    point_count_t chunkSize = std::min(m_chunkSize, count);
+    int nDims = dims.size();
+
+    std::map<std::string, pdalboost::any> buffers;
+
+    auto attrs = m_array->schema().attributes();
+
+    makeBuffer<double>(query, buffers, TILEDB_COORDS, chunkSize * dims.size());
+    auto spCoords = pdalboost::any_cast<std::shared_ptr<std::vector<double>>>(buffers[TILEDB_COORDS]);
+    query.set_coordinates(*spCoords);
+
+    if (!m_bbox.empty())
+    {
+        if (nDims == 2)
+            query.set_subarray({m_bbox.minx, m_bbox.minx, m_bbox.miny, m_bbox.maxy});
+        else
+            query.set_subarray({m_bbox.minx, m_bbox.minx, m_bbox.miny, m_bbox.maxy, m_bbox.minz, m_bbox.maxz});
+    }
+    else
+    {
+        // get extents
+        std::vector<double> subarray;
+        auto domain = m_array->non_empty_domain<double>();
+        for (const auto& kv : domain)
+        {
+            subarray.push_back(kv.second.first);
+            subarray.push_back(kv.second.second);
+        }
+        query.set_subarray(subarray);
+    }
+
+    for (const auto& a : attrs) {
+        std::string attrName = a.first;
+        Dimension::Id id = Dimension::id(attrName);
+        Dimension::Type t = Dimension::defaultType(id);
```

Comment:
defaultType() is a pretty-much useless call.  PDAL will convert from whatever your source type is, so you don't have to worry about this.  You can either a) just fetch your data as doubles and let PDAL convert it its internal storage type b) get the type of the dimension as it's stored in PDAL, which you can do by querying the layout (see dimTypes(), findDim() or findDimType()).